### PR TITLE
Add code feature extractor

### PR DESCRIPTION
**Pull request by Nokondi**
### Summary
- Added code feature extractor and decision tree classes
- Added code to static analysis function to redirect student project data to code feature extractor
- Added delegate and helper functions to extract features for assessing the learning goal 'Position - Elements and the Coordinate System' in unit 3 lesson 11
- Added decision tree to assess Position - Elements and the Coordinate System
- Added route test/assessment/cfe to test code feature extractor and decision tree pipeline using Position - Elements and the Coordinate System sample code
- Added additional unit test coverage for code_feature_extractor.py, decision_trees.py, and new code in label.py

### CODE REVIEW

File: lib/assessment/decision_trees.py
Comment (cearachew) [inline lib/assessment/decision_trees.py:1]: This looks good for now, but ideally I think we should try to have static decision trees stored in S3 with the prompts/etc and make this a class to read and evaluate static trees against the code features

File: lib/assessment/label.py [hunk 1]
```diff
@@ -43,6 +44,22 @@ def statically_label_student_work(self, rubric, student_code, student_id, exampl
...
+        for row in csv.DictReader(rubric.splitlines()):
+            if 'CFE' in row["Instructions"]:
+                cfe = CodeFeatures()
+                cfe.extract_features(student_code, row)
+                results["metadata"]["agent"] = ["code feature extractor", "static analysis"]
+                results["data"].append({"Label": cfe.assessment,
+                                        "Key Concept": row["Key Concept"],
+                                        "Observations": cfe.features,
+                                        "Reason": row[cfe.assessment] if cfe.assessment else ''
+                                        })
```
Comment (cearachew) [inline lib/assessment/label.py:59]: Do we have to modify all rubrics to have an "Instructions" row, or is this backwards compatible with rubrics that don't have this?

File: lib/assessment/label.py [hunk 2]
```diff
@@ -78,7 +95,7 @@ def ai_label_student_work(self, prompt, rubric, student_code, student_id, exampl
 
         return {
             'metadata': {
-                'agent': 'openai',
+                'agent': ['openai'],
```
Comment (davidsbailey) [inline lib/assessment/label.py:98]: is this change intentional?

**Review by davidsbailey — COMMENTED** [review]
great progress Mark! Let's discuss some of this feedback before you act on too much of it since I think some of it could potentially lead to some long follow-ups which we might not have time to invest in right now.

File: lib/assessment/label.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cearachew) [reply lib/assessment/label.py:59]: I like the term "static analysis" for anything that is not using AI (including detecting empty programs/ programs unchanged from starting code), and renaming this "cfe_label_student_work", or something similar

[the remark this answers is not in the document]
Comment (cearachew) [reply lib/assessment/label.py:59]: Also agree with Dave about not having the CFE enabled in the rubric itself via the "Instructions" column, since that could change the results from AI, unless there is code to remove the Instructions column prior to concatenating w/ the prompt and sending it to OpenAI

[the remark this answers is not in the document]
Comment (davidsbailey) [reply lib/assessment/label.py:59]: > Also agree with Dave about not having the CFE enabled in the rubric itself via the "Instructions" column, since that could change the results from AI, unless there is code to remove the Instructions column prior to concatenating w/ the prompt and sending it to OpenAI

To close the loop on one thing, the Instructions column cannot be removed for the purpose of CFE because it is part of our implementation for chain-of-thought (before and after swapping the Reason and Grade column). You'll find an Instructions column in every `standard_rubric.csv` file in S3.

**Review by davidsbailey — APPROVED** [review]
Looks fantastic Mark! I like the test coverage. Thanks for all the updates!

File: bin/cfe-assessment-test.rb
[the remark this answers is not in the document]
Comment (davidsbailey) [reply bin/cfe-assessment-test.rb:1]: > Thank you for adding this! It looks like this provides some nice end-to-end test coverage. a couple notes:
> 
> * To be clear, tests in this directory are not run as part of CI. I interpret the growing list of tests in this directory to indicate an unmet need for automated end-to-end testing capabilities.

https://codedotorg.atlassian.net/browse/AITT-497
